### PR TITLE
[IMP] odoo-print: automated printing settings

### PR DESCRIPTION
**Pull request by spenced**
Put all automated printing, both general printing strategies and
application-specific strategies, under a dedicated settings group.

Support execution of automated printing tests for derived models
providing specialised print strategies.

### CODE REVIEW

File: addons/print/tests/common.py [hunk 1]
```diff
@@ -36,7 +36,11 @@ def setUpClass(cls):
         # may be a derived class in a different module).
         module_file = sys.modules[cls.__module__].__file__
         module = get_resource_from_path(module_file)[0]
-        cls.files = pathlib.Path(get_resource_path(module, 'tests', 'files'))
+
+        try:
+            cls.files = pathlib.Path(get_resource_path(module, 'tests', 'files'))
+        except TypeError:
+            cls.files = None
```
Comment (mcb30) [inline addons/print/tests/common.py]: Should probably match https://github.com/unipartdigital/odoo-edi/commit/7c9e5dd867be919bd3f812e75f4cd89b1e42fc8f for consistency

File: addons/print/tests/test_action_print.py [hunk 2]
```diff
@@ -26,98 +27,132 @@ def model_id(cls, model):
...
+
+    cases = [
+        missing_strategy,
+        no_strategies,
+        missing_report,
+        missing_printer,
+        one_strategy,
+        two_strategies,
+        safety_catch,
+    ]
```
Comment (mcb30) [inline addons/print/tests/test_action_print.py]: Why the change to use an explicit list of cases instead of the standard search for method names beginning with "test"?  Having an explicit list makes it much easier to accidentally omit a test case from actually running.

File: addons/print/tests/test_action_print.py [hunk 3]
```diff
@@ -26,98 +27,132 @@ def model_id(cls, model):
...
+            try:
+                for func in body['cases']:
+                    yield func
+            except KeyError:
+                pass
+        def case(idx, func):
+            """Return a 2-tuple of (method name, func) for the `idx`th test."""
+            return ('test{:02d}_{:s}'.format(idx + 1, func.__name__), func)
+        body.update((case(idx, func) for (idx, func) in enumerate(cases())))
+        return super().__new__(mcs, name, bases, body)
```
Comment (mcb30) [inline addons/print/tests/test_action_print.py]: What is the purpose of this metaclass?

File: addons/print/tests/common.py (hunk 1 above)
Comment (spenced) [reply addons/print/tests/common.py]: Changed to be consistent.

File: addons/print/tests/test_action_print.py (hunk 2 above)
Comment (spenced) [reply addons/print/tests/test_action_print.py]: The unittest search probes all classes which are at the top-level of a test module, either because they are defined there or the class name is imported. Where a test class derives from a base test class in another module, the developer must be aware of this behaviour and must import the module for the base class, rather than the base class itself, to avoid the base class being executed (which is not normally desired). This is a trap which pushes developers to avoid using class inheritance in building out test behaviours.

The sequence of functions together with the accompanying metaclass allows a developer to explicitly state their intentions. Also, the numbering of test functions is automatically generated by the metaclass (across the derived class and all of its bases) meaning that a developer doesn't have to manually maintain the numbering sequence.

File: addons/print/tests/test_action_print.py (hunk 3 above)
Comment (spenced) [reply addons/print/tests/test_action_print.py]: (See previous comment.)

File: addons/print/tests/test_action_print.py (hunk 2 above)
Comment (mcb30) [reply addons/print/tests/test_action_print.py]: I really, **really** don't like the way that a test method has to be mentioned twice (once at the point of declaration, once in the "cases" list).  I'm pretty sure I would be guaranteed to forget to do this at least once.

Also: if a subclass wants to simply add a subcass-specific test case, how is it supposed to do this?  It's not immediately obvious to me.  Am I supposed to replicate the list of parent class cases (which will fail when the parent class is extended)?  Am I supposed to define a standard "def test*" method (which means that we now have multiple different mechanisms constructing the test case list)?  Am I supposed to perform some kind of explicit manipulation of the "cases" list (which isn't accessible at the time my subclass is constructed)?  Am I supposed to write my own metaclass?  None of these seem like good ideas compared to the standard approach.

The "import the module" workaround is admittedly ugly but is no less ugly, and has the advantage of failing much more gracefully: the failure mode for "import the module" is that test cases are executed twice, whereas the failure mode for "use metaclass and explicit cases list" is that test cases don't run at all.

How about instead adding a docstring comment to the parent class that mentions the problem with running test cases twice, and documents the correct way to fix it (by importing the module instead of the class)?  You'll need a docstring anyway to explain the non-standard "cases list" way of doing things, so documenting the alternative approach effectively costs nothing.

Comment (mcb30) [reply addons/print/tests/test_action_print.py]: Alternatively, I'd be happy with some metaclass-based approach that detects if the class is being detected only due to its being imported, and modifies its own behaviour to skip (or omit) its test cases in that circumstance.  I just think the current metaclass approach has too many subtle failure modes and consequently forces far too much thought onto the author of the subclass.

Comment (spenced) [reply addons/print/tests/test_action_print.py]: Moved into common and documented.

File: addons/print/tests/test_action_print.py (hunk 3 above)
Comment (spenced) [reply addons/print/tests/test_action_print.py]: Moved into common and documented.
